api/src/certification/complementary-certification/infrastructure/repositories/complementary-certification-badge-repository.js

```diff
@@ -33,4 +34,16 @@ const attach = async function ({ domainTransaction, complementaryCertificationBa
   }
 };
 
-export { getAllIdsByTargetProfileId, detachByIds, attach };
+const findAttachableBadgesByIds = async function ({ ids }) {
+  const badges = await knex
+    .from('badges')
+    .leftJoin('complementary-certification-badges', 'badges.id', 'complementary-certification-badges.badgeId')
+    .whereNull('complementary-certification-badges.id')
```

Comment:
Je trouve cette facon d'ecrire difficile à lire. Mais c'est ok sur le fonctionnement!

Reply:
Ce qui me viens à l'esprit c'est un where not exists

Reply:
Après analyze, l'usage de NOT EXISTS va effectivement utiliser l'algo de PostgreSQL Hash Anti-Join.

En parallèle j'ai fais des recherches, on a des résultats différents par type de BDD et par versions de la BDD ; mais il semble acquis que NOT EXISTS va globalement être soit équivalent au left join + is NULL, ou sinon être + performant.

Du coup je switche sur not exists :) Merci du retour !

Exemple de source : https://explainextended.com/2009/09/16/not-in-vs-not-exists-vs-left-join-is-null-postgresql/

Le query plan du NOT EXISTS qui montre le passage par l'algo Hash Anti Join ; 
```
QUERY PLAN                                                                                                                                |
------------------------------------------------------------------------------------------------------------------------------------------+
Hash Anti Join  (cost=10.68..21.25 rows=1 width=1622) (actual time=0.038..0.042 rows=3 loops=1)                                           |
  Hash Cond: (badges.id = "complementary-certification-badges".id)                                                                        |
  ->  Seq Scan on badges  (cost=0.00..10.55 rows=3 width=1622) (actual time=0.011..0.013 rows=3 loops=1)                                  |
        Filter: (id = ANY ('{601,700,61}'::integer[]))                                                                                    |
        Rows Removed by Filter: 21                                                                                                        |
  ->  Hash  (cost=10.30..10.30 rows=30 width=4) (actual time=0.012..0.012 rows=8 loops=1)                                                 |
        Buckets: 1024  Batches: 1  Memory Usage: 9kB                                                                                      |
        ->  Seq Scan on "complementary-certification-badges"  (cost=0.00..10.30 rows=30 width=4) (actual time=0.003..0.004 rows=8 loops=1)|
Planning Time: 0.114 ms                                                                                                                   |
Execution Time: 0.090 ms                                                                                                                  |
```

A titre de comparaison, le query plan du SELECT LEFT JOIN NULL, on voit que contrairement à l'article ci-dessus PostgreSQL ne passe **pas** par un Hash Anti join ; mais la version de PG est différente de l'article.
Pourquoi ? ça je sais pas 🤷🏻‍♂️ 

```
QUERY PLAN                                                                                                                             |
---------------------------------------------------------------------------------------------------------------------------------------+
Hash Right Join  (cost=10.59..20.98 rows=1 width=4238) (actual time=0.050..0.054 rows=3 loops=1)                                       |
  Hash Cond: ("complementary-certification-badges"."badgeId" = badges.id)                                                              |
  Filter: ("complementary-certification-badges".id IS NULL)                                                                            |
  ->  Seq Scan on "complementary-certification-badges"  (cost=0.00..10.30 rows=30 width=2616) (actual time=0.004..0.006 rows=8 loops=1)|
  ->  Hash  (cost=10.55..10.55 rows=3 width=1622) (actual time=0.030..0.031 rows=3 loops=1)                                            |
        Buckets: 1024  Batches: 1  Memory Usage: 9kB                                                                                   |
        ->  Seq Scan on badges  (cost=0.00..10.55 rows=3 width=1622) (actual time=0.012..0.016 rows=3 loops=1)                         |
              Filter: (id = ANY ('{601,700,61}'::integer[]))                                                                           |
              Rows Removed by Filter: 21                                                                                               |
Planning Time: 0.131 ms                                                                                                                |
Execution Time: 0.143 ms                                                                                                               |
```

Reply:
done !